### PR TITLE
General cheatcode

### PR DESCRIPTION
**Pull request by piotmag769**
Also removed changes that were made by us and were unnecessary - now we can clearly see the diff between this fork and original repo and plan on creating PRs to Cairo.

### CODE REVIEW

File: corelib/src/cheatcodes.cairo [hunk 1]
```diff
@@ -35,146 +24,64 @@ trait RevertedTransactionTrait {
...
+    loop {
+        match constructor_calldata.pop_front() {
+            Option::Some(value) => {
+                inputs.append(*value);
+            },
+            Option::None(_) => {
+                break ();
+            },
+        };
+    };
```
Comment (cptartur) [inline corelib/src/cheatcodes.cairo:61]: Why is this needed?

File: corelib/src/cheatcodes.cairo [hunk 2]
```diff
@@ -35,146 +24,64 @@ trait RevertedTransactionTrait {
...
+    let outputs = cheatcode::<'deploy'>(inputs.span());
+    let exit_code = *outputs[0];
+
+    if exit_code == 0 {
+        let result = *outputs[1];
+        Result::<felt252, RevertedTransaction>::Ok(result)
+    } else {
+        // TODO: feel free to change depending on the cheatcode::<'deploy'> low level implementation of error handling
+        let panic_data_len_felt = *outputs[1];
+        let panic_data_len = panic_data_len_felt.try_into().unwrap();
```
Comment (cptartur) [inline corelib/src/cheatcodes.cairo:72]: What is this conversion doing?

File: corelib/src/cheatcodes.cairo [hunk 3]
```diff
@@ -35,146 +24,64 @@ trait RevertedTransactionTrait {
...
+        let mut panic_data = array![];
+
+        let mut i = 2;
+        loop {
+            if panic_data_len + 2 == i {
+                break ();
             }
```
Comment (cptartur) [inline corelib/src/cheatcodes.cairo:79]: I don't like conditions like this, it's error prone. Can we somehow change it so we don't have to do `+2` in loop?

File: corelib/src/cheatcodes.cairo [hunk 4]
```diff
@@ -35,146 +24,64 @@ trait RevertedTransactionTrait {
...
-            prepared_contract = Option::Some(x);
-        },
-        Result::Err(x) => {
-            let mut panic_data = ArrayTrait::new();
-            panic_data.append(x);
-
-            return Result::<felt252, RevertedTransaction>::Err(RevertedTransaction { panic_data });
-        }
-    }
-    deploy(prepared_contract.unwrap())
```
Comment (cptartur) [inline corelib/src/cheatcodes.cairo:164]: Why were all other cheatcodes removed? They will be implemented at some point, it doesn't make sense to first remove and then add them again.

File: .gitignore [hunk 5]
```diff
@@ -3,4 +3,4 @@
 .idea/
 .spr.yml
 node_modules
-.DS_Store
+.DS_Store
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply .gitignore:6]: Can we automate this if its important?

[the remark this answers is not in the document]
Comment (cptartur) [reply .gitignore:6]: `cargo fmt` should do that already if you run it locally. There should be a tests for formatting in CI but it's not working on our fork now.

[the remark this answers is not in the document]
Comment (piotmag769) [reply .gitignore:6]: It deleted the newline on purpose, it is missing in original Cairo as well. I did it so it's easier to look at the [diff](https://github.com/starkware-libs/cairo/compare/main...software-mansion-labs:cairo:general-cheatcode)

File: corelib/src/cheatcodes.cairo [hunk 6]
```diff
@@ -35,146 +24,64 @@ trait RevertedTransactionTrait {
...
-    contract_address: felt252, function_name: felt252, calldata: @Array::<felt252>
-) -> Result::<(), RevertedTransaction> nopanic {
-    match invoke_impl(contract_address, function_name, calldata) {
-        Result::Ok(x) => Result::<(), RevertedTransaction>::Ok(x),
-        Result::Err(x) => Result::<(),
-        RevertedTransaction>::Err(RevertedTransaction { panic_data: x,  })
-    }
-}
+fn declare(contract: felt252) -> Result::<felt252, felt252> {
+    let span = cheatcode::<'declare'>(array![contract].span());
```
[the remark this answers is not in the document]
Comment (piotmag769) [reply corelib/src/cheatcodes.cairo:32]: It seems like it, I copied it from https://github.com/software-mansion-labs/cairo/blob/general-cheatcode/corelib/src/starknet/testing.cairo

File: corelib/src/cheatcodes.cairo (hunk 2 above)
Comment (piotmag769) [reply corelib/src/cheatcodes.cairo:72]: Converting from `felt252` to `usize`, you cannot access array's element by index using `felt252`.

File: corelib/src/cheatcodes.cairo (hunk 3 above)
Comment (piotmag769) [reply corelib/src/cheatcodes.cairo:79]: We can create an additional variable e.g. `let offset = i` before the loop if that suits you better

File: corelib/src/cheatcodes.cairo (hunk 4 above)
Comment (piotmag769) [reply corelib/src/cheatcodes.cairo:164]: They would have to reimplemented from scratch anyways, the code will differ (since arguments are passed in a different way, look the comment above about passing `Span<felt252>`). Leaving them here will result in unnecessary clutter. If one needs a reference to make sure the signature exposed to a user doesn't change, they can always check https://github.com/MaksymilianDemitraszek/cairo-protostar/

File: .gitignore (hunk 5 above)
[the remark this answers is not in the document]
Comment (piotmag769) [reply .gitignore:6]: Also the formatting now works on our fork (on this branch)

File: corelib/src/cheatcodes.cairo (hunk 4 above)
Comment (cptartur) [reply corelib/src/cheatcodes.cairo:164]: So you'd have to hunt these around the forks/prs/etc. I'd just leave them commented out here, and replaced them one by one.

Comment (piotmag769) [reply corelib/src/cheatcodes.cairo:164]: Discussed offline, will leave some of them commented out.

**Review by karol-bisztyga — APPROVED** [review]

File: corelib/src/cheatcodes.cairo [hunk 7]
```diff
@@ -35,146 +24,141 @@ trait RevertedTransactionTrait {
...
-fn call(
-    contract: felt252, function_name: felt252, calldata: @Array::<felt252>
-) -> Result::<Array<felt252>, RevertedTransaction> nopanic {
-    match call_impl(contract, function_name, calldata) {
-        Result::Ok(x) => Result::<Array<felt252>, RevertedTransaction>::Ok(x),
-        Result::Err(x) => Result::<Array<felt252>,
-        RevertedTransaction>::Err(RevertedTransaction { panic_data: x,  })
-    }
-}
+// Old code commented out for future reference - signatures etc.
```
Comment (karol-bisztyga) [inline corelib/src/cheatcodes.cairo:90]: This doesn't make sense to me, I remember we've talked about this when migrating to the new arch. We use git so we don't have to leave something like this, please remove.

**Review by karol-bisztyga — CHANGES_REQUESTED** [review]

File: corelib/src/cheatcodes.cairo [hunk 8]
```diff
@@ -1,28 +1,17 @@
...
-extern fn stop_prank(target_contract_address: felt252) -> Result::<(), felt252> nopanic;
-
-extern fn declare(contract: felt252) -> Result::<felt252, felt252> nopanic;
-
-extern fn declare_cairo0(contract: felt252) -> Result::<felt252, felt252> nopanic;
+#[derive(Drop, Clone)]
+struct PreparedContract {
+    contract_address: felt252,
+    class_hash: felt252,
+    constructor_calldata: Array::<felt252>,
```
Comment (cptartur) [inline corelib/src/cheatcodes.cairo:13]: It should still use `@Array::<felt252>`

File: corelib/src/cheatcodes.cairo [hunk 9]
```diff
@@ -35,146 +24,141 @@ trait RevertedTransactionTrait {
...
+//     block_timestamp: felt252, target_contract_address: felt252
+// ) -> Result::<(), felt252> nopanic;
+
+// extern fn stop_warp(target_contract_address: felt252) -> Result::<(), felt252> nopanic;
+
+// extern fn start_prank(
+//     caller_address: felt252, target_contract_address: felt252
+// ) -> Result::<(), felt252> nopanic;
+
+// extern fn stop_prank(target_contract_address: felt252) -> Result::<(), felt252> nopanic;
```
Comment (cptartur) [inline corelib/src/cheatcodes.cairo:164]: Newline

**Review by cptartur — CHANGES_REQUESTED** [review]

File: corelib/src/cheatcodes.cairo (hunk 7 above)
Comment (piotmag769) [reply corelib/src/cheatcodes.cairo:90]: I support removing it as well, I don't think we will have to reference any PR, you can just use `git checkout`. Please discuss it offline and come back when you reach an agreement

Comment (karol-bisztyga) [reply corelib/src/cheatcodes.cairo:90]: Discussed offline, nobody managed to convince the other one so I guess the decision's yours @piotmag769

File: corelib/src/cheatcodes.cairo [hunk 10]
```diff
@@ -35,146 +24,141 @@ trait RevertedTransactionTrait {
...
+    };
+
+    let outputs = cheatcode::<'deploy'>(inputs.span());
+    let exit_code = *outputs[0];
+
+    if exit_code == 0 {
+        let result = *outputs[1];
+        Result::<felt252, RevertedTransaction>::Ok(result)
+    } else {
+        // TODO: feel free to change depending on the cheatcode::<'deploy'> low level implementation of error handling
```
[the remark this answers is not in the document]
Comment (piotmag769) [reply corelib/src/cheatcodes.cairo:70]: There is already a PR for that tho, it is just a draft to help with implementing it 
https://github.com/software-mansion/protostar/pull/2157/files

File: corelib/src/cheatcodes.cairo (hunk 7 above)
Comment (piotmag769) [reply corelib/src/cheatcodes.cairo:90]: Will remove it then

File: corelib/src/cheatcodes.cairo (hunk 10 above)
[the remark this answers is not in the document]
Comment (piotmag769) [reply corelib/src/cheatcodes.cairo:70]: There is already a PR that should be rebased on top of this one after merging. It is just a draft for error handling for @Arcticae

**Review by karol-bisztyga — APPROVED** [review]

**Review by cptartur — APPROVED** [review]
